File: frontend/app/src/types/mpf.ts
export interface MpfFundQuote {
  fund_cd: string
  fund_class: string
  description_en: string
  description_zh: string
  category: string
  nav: number
  valuation_date: string
}

export interface MpfAllocationTarget {
  fund_cd: string
  fund_class: string
  description_en: string
  description_zh: string
  percentage: number
}

export interface MpfPortfolio {
  id: string
  entity_id: string
  entity_name?: string
  name: string
  scheme: string
  currency: string
  target_allocation: MpfAllocationTarget[]
  created_at: string
}

export interface MpfContributionLineItem {
  fund_cd: string
  fund_class: string
  description_en: string
  description_zh: string
  percentage: number
  amount: number
  nav: number
  units: number
}

export interface MpfContribution {
  id: string
  portfolio_id: string
  date: string
  total_amount: number
  currency: string
  note?: string
  line_items: MpfContributionLineItem[]
}

export interface MpfHoldingFund {
  fund_cd: string
  fund_class: string
  description_en: string
  description_zh: string
  units: number
  nav: number
  market_value: number
}

export interface MpfPortfolioSummary {
  portfolio: MpfPortfolio
  holdings: MpfHoldingFund[]
  total_contributed: number
  total_market_value: number
}

export interface CreateMpfPortfolioPayload {
  entity_id: string
  name: string
  scheme: string
  currency: string
  target_allocation: MpfAllocationTarget[]
}

export interface UpdateMpfPortfolioPayload {
  name: string
  target_allocation: MpfAllocationTarget[]
}

export interface RecordMpfContributionPayload {
  date: string
  total_amount: number
  note?: string
}

File: finanze/domain/mpf.py
from datetime import date, datetime
from typing import Optional
from uuid import UUID

from domain.dezimal import Dezimal
from pydantic.dataclasses import dataclass


@dataclass(kw_only=True)
class MpfFundQuote:
    fund_cd: str
    fund_class: str
    description_en: str
    description_zh: str
    category: str
    nav: Dezimal
    valuation_date: date


@dataclass(kw_only=True)
class MpfAllocationTarget:
    fund_cd: str
    fund_class: str
    description_en: str
    description_zh: str
    percentage: Dezimal


@dataclass(kw_only=True)
class MpfPortfolio:
    id: UUID
    entity_id: UUID
    entity_name: Optional[str] = None
    name: str
    scheme: str
    currency: str
    target_allocation: list[MpfAllocationTarget]
    created_at: datetime


@dataclass(kw_only=True)
class MpfContributionLineItem:
    fund_cd: str
    fund_class: str
    description_en: str
    description_zh: str
    percentage: Dezimal
    amount: Dezimal
    nav: Dezimal
    units: Dezimal


@dataclass(kw_only=True)
class MpfContribution:
    id: UUID
    portfolio_id: UUID
    date: datetime
    total_amount: Dezimal
    currency: str
    note: Optional[str] = None
    line_items: list[MpfContributionLineItem]


@dataclass(kw_only=True)
class MpfHoldingFund:
    fund_cd: str
    fund_class: str
    description_en: str
    description_zh: str
    units: Dezimal
    nav: Dezimal
    market_value: Dezimal


@dataclass(kw_only=True)
class MpfPortfolioSummary:
    portfolio: MpfPortfolio
    holdings: list[MpfHoldingFund]
    total_contributed: Dezimal
    total_market_value: Dezimal


@dataclass(kw_only=True)
class CreateMpfPortfolioRequest:
    entity_id: UUID
    name: str
    scheme: str
    currency: str
    target_allocation: list[MpfAllocationTarget]


@dataclass(kw_only=True)
class UpdateMpfPortfolioRequest:
    id: UUID
    name: str
    target_allocation: list[MpfAllocationTarget]


@dataclass(kw_only=True)
class RecordMpfContributionRequest:
    portfolio_id: UUID
    date: datetime
    total_amount: Dezimal
    note: Optional[str] = None
